### ISSUE TITLE
Output changes after Prefix warmup.

### ISSUE THREAD

**Issue by DouHappy** [vllm-project/vllm#2129]
@caoshiyi

I think ouput should be equal when setting prefix for the first time and after. But my test shows that output will changes after you first set this prefix. In addention, output of without prefix will also affected when batchsize is too big(seems when requests can't be push into scheduler.running at once)

I have set temperature=0 to obtain a fixed output.

Model: llama-7b,
prompt:
```
You are a helpful assistant in recongnizes the content of tables in markdown format. Here is a table as fellows. You need to answer my question about the table.\n# Table\n|Opening|Opening|Sl. No.|Film|Cast|Director|Music Director|Notes|\n|----|----|----|----|----|----|----|----|\n|J A N|9|1|Agni Pushpam|Jayabharathi, Kamalahasan|Jeassy|M. K. Arjunan||\n|J A N|16|2|Priyamvada|Mohan Sharma, Lakshmi, KPAC Lalitha|K. S. Sethumadhavan|V. Dakshinamoorthy||\n|J A N|23|3|Yakshagaanam|Madhu, Sheela|Sheela|M. S. Viswanathan||\n|J A N|30|4|Paalkkadal|Sheela, Sharada|T. K. Prasad|A. T. Ummer||\n|F E B|5|5|Amma|Madhu, Srividya|M. Krishnan Nair|M. K. Arjunan||\n|F E B|13|6|Appooppan|Thikkurissi Sukumaran Nair, Kamal Haasan|P. Bhaskaran|M. S. Baburaj||\n|F E B|20|7|Srishti|Chowalloor Krishnankutty, Ravi Alummoodu|K. T. Muhammad|M. S. Baburaj||\n|F E B|20|8|Vanadevatha|Prem Nazir, Madhubala|Yusufali Kechery|G. Devarajan||\n|F E B|27|9|Samasya|Madhu, Kamalahaasan|K. Thankappan|Shyam||\n|F E B|27|10|Yudhabhoomi|K. P. Ummer, Vidhubala|Crossbelt Mani|R. K. Shekhar||\n|M A R|5|11|Seemantha Puthran|Prem Nazir, Jayabharathi|A. B. Raj|M. K. Arjunan||\n|M A R|12|12|Swapnadanam|Rani Chandra, Dr. Mohandas|K. G. George|Bhaskar Chandavarkar||\n|M A R|19|13|Thulavarsham|Prem Nazir, sreedevi, Sudheer|N. Sankaran Nair|V. Dakshinamoorthy||\n|M A R|20|14|Aruthu|Kaviyoor Ponnamma, Kamalahasan|Ravi|G. Devarajan||\n|M A R|26|15|Swimming Pool|Kamal Haasan, M. G. Soman|J. Sasikumar|M. K. Arjunan||\n\n# Question\nWhat' s the content in the (1,1) cells\n
```
Prifix Length: 592
warmup output:
```
# Explanation\n\n
# 1. Opening\n
# 2. Opening\n
# 3. Sl. No.\n
# 4. Film\n
# 5. Cast\n
# 6. Director\n
# 7. Music Director\n
# 8. Notes\n
# 9. Sl. No.\n
# 10. Film\n
# 
```
after warmup:
```

# Explanation\n\n
# Table\n|Opening|Opening|Sl. No.|Film|Cast|Director|Music Director|Notes|\n|----|----|----|----|----|----|----|----|\n|J A N|9|1|Agni Pushpam|Jayabharathi, Kamalahasan|Jeassy|M. K. Arjunan
```

My test script(imported from jupyter):
```python
# %%
# generate test prompt
test_table = "|Opening|Opening|Sl. No.|Film|Cast|Director|Music Director|Notes|\n|----|----|----|----|----|----|----|----|\n|J A N|9|1|Agni Pushpam|Jayabharathi, Kamalahasan|Jeassy|M. K. Arjunan||\n|J A N|16|2|Priyamvada|Mohan Sharma, Lakshmi, KPAC Lalitha|K. S. Sethumadhavan|V. Dakshinamoorthy||\n|J A N|23|3|Yakshagaanam|Madhu, Sheela|Sheela|M. S. Viswanathan||\n|J A N|30|4|Paalkkadal|Sheela, Sharada|T. K. Prasad|A. T. Ummer||\n|F E B|5|5|Amma|Madhu, Srividya|M. Krishnan Nair|M. K. Arjunan||\n|F E B|13|6|Appooppan|Thikkurissi Sukumaran Nair, Kamal Haasan|P. Bhaskaran|M. S. Baburaj||\n|F E B|20|7|Srishti|Chowalloor Krishnankutty, Ravi Alummoodu|K. T. Muhammad|M. S. Baburaj||\n|F E B|20|8|Vanadevatha|Prem Nazir, Madhubala|Yusufali Kechery|G. Devarajan||\n|F E B|27|9|Samasya|Madhu, Kamalahaasan|K. Thankappan|Shyam||\n|F E B|27|10|Yudhabhoomi|K. P. Ummer, Vidhubala|Crossbelt Mani|R. K. Shekhar||\n|M A R|5|11|Seemantha Puthran|Prem Nazir, Jayabharathi|A. B. Raj|M. K. Arjunan||\n|M A R|12|12|Swapnadanam|Rani Chandra, Dr. Mohandas|K. G. George|Bhaskar Chandavarkar||\n|M A R|19|13|Thulavarsham|Prem Nazir, sreedevi, Sudheer|N. Sankaran Nair|V. Dakshinamoorthy||\n|M A R|20|14|Aruthu|Kaviyoor Ponnamma, Kamalahasan|Ravi|G. Devarajan||\n|M A R|26|15|Swimming Pool|Kamal Haasan, M. G. Soman|J. Sasikumar|M. K. Arjunan||"

prompt_template = '''\
You are a helpful assistant in recongnizes the content of tables in markdown format. Here is a table as fellows. You need to answer my question about the table.
# Table
{}

# Question
What' s the content in the ({},{}) cells
'''
with open('prompt.txt', 'w') as outer:
    for row in range(50):
        for column in range(10):
            tmp_str = prompt_template.format(test_table, row + 1, column + 1)
            tmp_str
            # outer.write(f"{tmp_str}")
            print(tmp_str.replace("\n", "\\n"), file=outer)

# %%
import time
import datetime
import os

from vllm import LLM
from vllm import SamplingParams

import torch

def test_prefix(llm = None, sampling_params=None, prompts=None, prefix_len=None, save_file=None, detile=True):
    if isinstance(prefix_len, int):
        prefix_len = [prefix_len]
        assert len(prompts) % len(prefix_len) == 0, f"len of prompts must be multiple of len of prefix_len"
    print("------start generating------")
    start_time = time.time()
    # whether use Prefix
    if prefix_len != None:
        # start inference
        outputs = llm.generate(prompts, sampling_params=sampling_params, prefix_pos=prefix_len * (len(prompts) // len(prefix_len)))
    else:
        outputs = llm.generate(prompts, sampling_params=sampling_params)

    end_time = time.time()
    print(f"cost time {end_time - start_time}")

    if save_file != None:
        print("saving output......")
        for index, output in enumerate(outputs):
            if detile == True:
                print(output, file=save_file)
            else:
                print(output.outputs[0].text, file=save_file)
                # print({'id': index, 'output':output.outputs[0].text}, file=save_file)
        print(f"output saved in {save_file.name} {datetime.datetime.now()}")

# %%
# set gpus
os.environ['CUDA_VISIBLE_DEVICES']="0"
# init model and sampling parames
tensor_parallel_size = len(os.getenv('CUDA_VISIBLE_DEVICES').split(','))
# set baichuan model
# model = "/data/images/llms/models--baichuan-inc--Baichuan2-13B-Chat"
# model = "/data/images/llms/huggingface/hub/models--THUDM--chatglm3-6b/snapshots/e46a14881eae613281abbd266ee918e93a56018f/"
model = "/data/images/llms/huggingface/hub/models--huggyllama--llama-7b/snapshots/8416d3fefb0cb3ff5775a7b13c1692d10ff1aa16"
# Create an LLM.
llm = LLM(model=model, tokenizer_mode='auto', trust_remote_code=True, tensor_parallel_size=tensor_parallel_size)

# %%
# get prompts
prompts = []
with open("prompt.txt", 'r') as reader:
    prompts = reader.readlines()

sampling_params = SamplingParams(temperature=0, max_tokens=100)

with open("output.txt", 'w') as f:
    test_prefix(llm=llm,
                prompts=prompts[:1],
                prefix_len=[592],
                # prefix_len=None,
                save_file=f,
                sampling_params=sampling_params,
                detile=False,
                )


with open("output2.txt", 'w') as f:
    test_prefix(llm=llm,
                prompts=prompts[:1],
                prefix_len=[592],
                # prefix_len=None,
                save_file=f,
                sampling_params=sampling_params,
                detile=False,
                )
```

Some experiments :
warmup[:100] means in the 79th line of my script, warmup by 100 prompts with the same prefix.

| model    | batchsize | prefix | notes        | md5sum                           |
| -------- | --------- | ------ | ------------ | -------------------------------- |
| llama-7b | 100       |        | run directly | c866ce73289dda6757139ea6221e932c |
| llama-7b | 100       |        | warmup[:100] | 894435281f1a0437b11648e6bf26f1ca |
| llama-7b | 100       | 592    | run directly | 194dd58b77fc5f5bc93f9335eb281303 |
| llama-7b | 100       | 592    | warmup[:100] | 1b592c147f80c10f704f226239b54095 |

| model | batchsize | prefix | notes | md5sum |
| --- | --- | --- | --- | --- |
| llama-7b | 50 |  | run directly | d9dbeeb4056ae6fbe7a482f12161adcc |
| llama-7b | 50 |  | warmup[:50] | d9dbeeb4056ae6fbe7a482f12161adcc |
| llama-7b | 50 |  | warmup[:100] | d9dbeeb4056ae6fbe7a482f12161adcc |
| llama-7b | 50 | 592 | run directly | f6308f6440231f96dbc20040812c1d66 |
| llama-7b | 50 | 592 | warmup[:50] | f6308f6440231f96dbc20040812c1d66 |
| llama-7b | 50 | 592 | warmup[:100] | f6308f6440231f96dbc20040812c1d66 |

| model | batchsize | prefix | notes | md5sum |
| --- | --- | --- | --- | --- |
| llama-7b | 1 |  | run directly | 699f4948c0d54ea92ef4564f61e708de |
| llama-7b | 1 |  | warmup[:1] | 699f4948c0d54ea92ef4564f61e708de |
| llama-7b | 1 | 592 | run directly | 699f4948c0d54ea92ef4564f61e708de |
| llama-7b | 1 | 592 | warmup[:1] | 1b0955379c444896047663cdad1c9b7d |

Hope This helps.

**Comment by caoshiyi** [comment]
Thanks for the issue. I will look into this.

**Comment by caoshiyi** [comment]
@DouHappy I used your script and replaced line 72-73 with the prompt you provided:
```
You are a helpful assistant in recongnizes the content of tables in markdown format. Here is a table as fellows. You need to answer my question about the table.\n# Table\n|Opening|Opening|Sl. No.|Film|Cast|Director|Music Director|Notes|\n|----|----|----|----|----|----|----|----|\n|J A N|9|1|Agni Pushpam|Jayabharathi, Kamalahasan|Jeassy|M. K. Arjunan||\n|J A N|16|2|Priyamvada|Mohan Sharma, Lakshmi, KPAC Lalitha|K. S. Sethumadhavan|V. Dakshinamoorthy||\n|J A N|23|3|Yakshagaanam|Madhu, Sheela|Sheela|M. S. Viswanathan||\n|J A N|30|4|Paalkkadal|Sheela, Sharada|T. K. Prasad|A. T. Ummer||\n|F E B|5|5|Amma|Madhu, Srividya|M. Krishnan Nair|M. K. Arjunan||\n|F E B|13|6|Appooppan|Thikkurissi Sukumaran Nair, Kamal Haasan|P. Bhaskaran|M. S. Baburaj||\n|F E B|20|7|Srishti|Chowalloor Krishnankutty, Ravi Alummoodu|K. T. Muhammad|M. S. Baburaj||\n|F E B|20|8|Vanadevatha|Prem Nazir, Madhubala|Yusufali Kechery|G. Devarajan||\n|F E B|27|9|Samasya|Madhu, Kamalahaasan|K. Thankappan|Shyam||\n|F E B|27|10|Yudhabhoomi|K. P. Ummer, Vidhubala|Crossbelt Mani|R. K. Shekhar||\n|M A R|5|11|Seemantha Puthran|Prem Nazir, Jayabharathi|A. B. Raj|M. K. Arjunan||\n|M A R|12|12|Swapnadanam|Rani Chandra, Dr. Mohandas|K. G. George|Bhaskar Chandavarkar||\n|M A R|19|13|Thulavarsham|Prem Nazir, sreedevi, Sudheer|N. Sankaran Nair|V. Dakshinamoorthy||\n|M A R|20|14|Aruthu|Kaviyoor Ponnamma, Kamalahasan|Ravi|G. Devarajan||\n|M A R|26|15|Swimming Pool|Kamal Haasan, M. G. Soman|J. Sasikumar|M. K. Arjunan||\n\n# Question\nWhat' s the content in the (1,1) cells\n
```
and my output for both with and without prefix (bs=1,50,1000, w/&w/o warmup) is: 
```
Answer:

\begin{code}
# Table
|Opening|Opening|Sl. No.|Film|Cast|Director|Music Director|Notes|
|----|----|----|----|----|----|----|----|
|J A N|9|1|Agni Pushpam|Jayabharathi, Kamalahasan|Jeassy|M. K. Arjunan
```
I used the model `huggyllama/llama-7b`.

**Comment by jadielam** [comment]
I have been searching the whole code base, trying to find the parameter `prefix_pos` that you pass to the `llm.generate` method, but I cannot find it. Are you using a branch other than the main branch for this?
